### ISSUE TITLE
errors updating build config last triggered image id when tailing build logs

### ISSUE THREAD

**Issue by rcernich** [openshift/origin#1876]
I've noticed that builds get continuously scheduled when tailing a running build's logs (i.e. build-logs).  During this time, os will schedule another build (and another and another).  If I don't tail logs, no other builds are scheduled.  There are errors in the openshift-master logs about updating buildConfig with LastTriggeredImageID.

**Comment by bparees** [comment]
@rcernich please go and and update this issue with loglevel=5 logs and the output of osc get imagestreams, osc get builds, osc get buildconfigs  if/when you reproduce this again with the latest openshift (where I added the additional level of error logging so we can see why the etcd update is failing)

**Comment by rcernich** [comment]
Some more info:
```
May 22 23:57:50 ce-openshift-rhel-master openshift[592]: E0522 23:57:50.624238     592 factory.go:155] fatal error handling ImageStream change: Error updating buildConfig jws-app with new LastTriggeredImageID, root error was: buildconfig "jws-app" cannot be updated: the resource was updated to 235485
```

**Comment by rcernich** [comment]
This output is interesting:
```
curl --cert /openshift/openshift.local.config/master/master.etcd-client.crt --cacert /openshift/openshift.local.config/master/ca.crt --key /openshift/openshift.local.config/master/master.etcd-client.key -k https://localhost:4001/v2/keys/openshift.io
{"action":"get","node":{"key":"/openshift.io","dir":true,"nodes":[{"key":"/openshift.io/templates","dir":true,"modifiedIndex":43,"createdIndex":43},{"key":"/openshift.io/authorization","dir":true,"modifiedIndex":11,"createdIndex":11},{"key":"/openshift.io/deploymentconfigs","dir":true,"modifiedIndex":180,"createdIndex":180},{"key":"/openshift.io/images","dir":true,"modifiedIndex":856,"createdIndex":856},{"key":"/openshift.io/oauth","dir":true,"modifiedIndex":4,"createdIndex":4},{"key":"/openshift.io/registry","dir":true,"modifiedIndex":181027,"createdIndex":181027},{"key":"/openshift.io/users","dir":true,"modifiedIndex":486,"createdIndex":486},{"key":"/openshift.io/buildconfigs","dir":true,"modifiedIndex":1146,"createdIndex":1146},{"key":"/openshift.io/builds","dir":true,"modifiedIndex":1177,"createdIndex":1177},{"key":"/openshift.io/imagerepositories","dir":true,"modifiedIndex":847,"createdIndex":847},{"key":"/openshift.io/routes","dir":true,"modifiedIndex":197,"createdIndex":197},{"key":"/openshift.io/useridentities","dir":true,"modifiedIndex":487,"createdIndex":487}],"modifiedIndex":4,"createdIndex":4}}
```
I don't see buildConfigs in there.

**Comment by rcernich** [comment]
Sorry, apparently, I can't read.  buildconfigs is there.

**Comment by bparees** [comment]
@csrwng if you could look into this while i'm out i'd appreciate it.  the failure mode is pretty basic (image_change_controller.go is trying to update the buildConfig around line 129), but the meaning of the etcd error is not very obvious to me.

**Comment by rcernich** [comment]
One thing that is odd, is that it seems like the update succeeds if you curl the key for the build config.

**Comment by rcernich** [comment]
I think I can confirm that after you issue a get request for the buildconfigs key using curl, the spawning of builds stops.  In this last case, the request was for buildconfigs/project.

**Comment by ncdc** [comment]
We've updated the code a little bit. It's possible to still see a similar error (`glog.V(4).Infof("Failed to update BuildConfig %s/%s so no Build will be created", bc.Namespace, bc.Name)`) in the event that 2 build image change controllers both try to handle the same event. In the future, we'll update the way we process image changes and builds to try to avoid these log messages (with a fix for #2604), but they're essentially harmless.

Please reopen if you think there's still an issue - thanks.
